pallets/block-rewards/src/lib.rs

```diff
@@ -0,0 +1,499 @@
+}
+
+/// Type that contains the associated data of an session
+#[derive(Encode, Decode, TypeInfo, MaxEncodedLen, PartialEq, Eq, RuntimeDebugNoBound)]
+#[scale_info(skip_type_params(T))]
+pub struct SessionData<T: Config> {
+	/// Amount of rewards per session for a single collator.
+	pub(crate) collator_reward: T::Balance,
+	/// Total amount of rewards per session
+	/// NOTE: Is ensured to be at least collator_reward * collator_count.
+	pub(crate) total_reward: T::Balance,
+	/// Number of current collators.
+	/// NOTE: Updated automatically and thus not adjustable via extrinsic.
+	pub collator_count: u32,
+}
+
+impl<T: Config> Default for SessionData<T> {
+	fn default() -> Self {
+		Self {
+			collator_reward: T::Balance::zero(),
+			total_reward: T::Balance::zero(),
+			collator_count: 0,
+		}
+	}
+}
+
+/// Type that contains the pending update.
+#[derive(
+	PartialEq, Clone, DefaultNoBound, Encode, Decode, TypeInfo, MaxEncodedLen, RuntimeDebugNoBound,
+)]
+#[scale_info(skip_type_params(T))]
+pub struct SessionChanges<T: Config> {
+	pub collators: CollatorChanges<T>,
+	pub collator_count: Option<u32>,
+	collator_reward: Option<T::Balance>,
+	total_reward: Option<T::Balance>,
+}
+
+pub(crate) type DomainIdOf<T> = <<T as Config>::Domain as TypedGet>::Type;
+pub(crate) type NegativeImbalanceOf<T> = <<T as Config>::Currency as CurrencyT<
+	<T as frame_system::Config>::AccountId,
+>>::NegativeImbalance;
+
+#[frame_support::pallet]
+pub mod pallet {
+
+	use super::*;
+
+	pub const STORAGE_VERSION: StorageVersion = StorageVersion::new(1);
+
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RuntimeEvent: From<Event<Self>> + IsType<<Self as frame_system::Config>::RuntimeEvent>;
+
+		/// Required origin for admin purposes for configuring groups and currencies.
+		type AdminOrigin: EnsureOrigin<Self::RuntimeOrigin>;
+
+		/// Type used to handle balances.
+		type Balance: Balance
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ Into<<<Self as Config>::Currency as CurrencyT<Self::AccountId>>::Balance>
+			+ MaybeSerializeDeserialize;
+
+		/// Domain identification used by this pallet
+		type Domain: TypedGet;
+
+		/// Type used to handle group weights.
+		type Weight: Parameter + MaxEncodedLen + EnsureAdd + Unsigned + FixedPointOperand + Default;
+
+		/// The reward system used.
+		type Rewards: GroupRewards<Balance = Self::Balance, GroupId = u32>
+			+ AccountRewards<
+				Self::AccountId,
+				Balance = Self::Balance,
+				CurrencyId = (DomainIdOf<Self>, CfgCurrencyId),
+			> + CurrencyGroupChange<GroupId = u32, CurrencyId = (DomainIdOf<Self>, CfgCurrencyId)>;
+
+		/// Type used to handle currency minting and burning for collators.
+		type Currency: Mutate<Self::AccountId, AssetId = CfgCurrencyId, Balance = Self::Balance>
+			+ CurrencyT<Self::AccountId>;
+
+		/// The identifier of the artificial block rewards currency which is minted and burned for collators.
+		#[pallet::constant]
+		type StakeCurrencyId: Get<CfgCurrencyId>;
```

Comment:
I mean, the pallet can works without knowing that the `CurrencyId` is exactly a `CfgCurrencyId`

Reply:
Yes, we have to add an `AssetId` associated type then. Will do this tomorrow! Good catch.

Reply:
No hurry! Maybe a `CurrencyIdOf<T>` using the `CurrencyId` of `Rewards` traits or The `Currency::AssetId` saves you form another Config parameter

Reply:
ecd49deb50e1149617f4220be2589a9095c7f32b

Reply:
Since we need to specify the `CurrencyId` for the `Rewards` trait, I don't see how we get around adding a `CurrencyId` parameter to the `pallet_block_rewards::Config`.